clients/src/test/java/org/apache/kafka/clients/consumer/internals/ConsumerInterceptorsTest.java

```diff
@@ -0,0 +1,164 @@
+            return new ConsumerRecords<K, V>(recordMap);
+        }
+
+        @Override
+        public void onCommit(Map<TopicPartition, OffsetAndMetadata> offsets) {
+            onCommitCount++;
+            if (throwExceptionOnCommit)
+                throw new KafkaException("Injected exception in FilterConsumerInterceptor.onCommit.");
+        }
+
+        @Override
+        public void close() {
+        }
+
+        // if 'on' is true, onConsume will always throw an exception
+        public void injectOnConsumeError(boolean on) {
+            throwExceptionOnConsume = on;
+        }
+
+        // if 'on' is true, onConsume will always throw an exception
+        public void injectOnCommitError(boolean on) {
+            throwExceptionOnCommit = on;
+        }
+    }
+
+    @Test
+    public void testMutableInterceptorChain() {
+        List<ConsumerInterceptor<Integer, Integer>>  interceptorList = new ArrayList<>();
+        // we are testing two different interceptors by configuring the same interceptor differently, which is not
+        // how it would be done in KafkaConsumer, but ok for testing interceptor callbacks
+        FilterConsumerInterceptor<Integer, Integer> interceptor1 = new FilterConsumerInterceptor<>(filterPartition1);
+        FilterConsumerInterceptor<Integer, Integer> interceptor2 = new FilterConsumerInterceptor<>(filterPartition2);
+        interceptorList.add(interceptor1);
+        interceptorList.add(interceptor2);
+        ConsumerInterceptors<Integer, Integer> interceptors = new ConsumerInterceptors<>(interceptorList);
+
+        // verify that onConsumer mutates ConsumerRecords
+        Map<TopicPartition, List<ConsumerRecord<Integer, Integer>>> records = new HashMap<>();
+        List<ConsumerRecord<Integer, Integer>> list1 = new ArrayList<>();
+        list1.add(consumerRecord);
+        List<ConsumerRecord<Integer, Integer>> list2 = new ArrayList<>();
+        list2.add(new ConsumerRecord<>(filterTopicPart1.topic(), filterTopicPart1.partition(), 0, 1, 1));
+        List<ConsumerRecord<Integer, Integer>> list3 = new ArrayList<>();
+        list3.add(new ConsumerRecord<>(filterTopicPart2.topic(), filterTopicPart2.partition(), 0, 1, 1));
+        records.put(tp, list1);
+        records.put(filterTopicPart1, list2);
+        records.put(filterTopicPart2, list3);
+        ConsumerRecords<Integer, Integer> consumerRecords = new ConsumerRecords<>(records);
+        ConsumerRecords<Integer, Integer> interceptedRecords = interceptors.onConsume(consumerRecords);
+        assertEquals(1, interceptedRecords.count());
+        assertTrue(interceptedRecords.partitions().contains(tp));
+        assertFalse(interceptedRecords.partitions().contains(filterTopicPart1));
+        assertFalse(interceptedRecords.partitions().contains(filterTopicPart2));
+        assertEquals(2, onConsumeCount);
+
+        // verify that even if one of the intermediate interceptors throws an exception, all interceptors' onConsume are called
```

Comment:
Yes, let me separate into two tests: one for all onConsume() and one for all onCommit() calls.